[C API] Add getters and setters for fast-math flags on relevant instructions

Benjins opened the pull request:
These flags are usable on floating point arithmetic, as well as call, select, and phi instructions whose resulting type is floating point, or a vector of, or an array of, a valid type. Whether or not the flags are valid for a given instruction can be checked with the new LLVMGetCanUseFastMathFlags function

These are exposed using a new LLVMFastMathFlags type, which is an alias for unsigned. An anonymous enum defines the bit values for it

Tests are added in echo.ll for select/phil/call, and the floating point types in the new float_ops.ll bindings test

Select and the floating point arithmetic instructions were not implemented in llvm-c-test/echo.cpp, so they were added as well

Benjins committed:
[C API] Add getters and setters for fast-math flags on relevant instructions

These flags are usable on floating point arithmetic, as well as call, select,
and phi instructions whose resulting type is floating point, or a vector of,
or an array of, a valid type. Whether or not the flags are valid for a given
instruction can be checked with the new LLVMGetCanUseFastMathFlags function

These are exposed using a new LLVMFastMathFlags type, which is an alias for
unsigned. An anonymous enum defines the bit values for it

Tests are added in echo.ll for select/phil/call, and the floating point types
in the new float_ops.ll bindings test

Select and the floating point arithmetic instructions were not implemented in
llvm-c-test/echo.cpp, so they were added as well

Benjins committed:
Fix typo in docs for LLVMGetCanUseFastMathFlags

nikic commented on llvm/include/llvm-c/Core.h:
```diff
@@ -483,6 +483,26 @@ typedef enum {
+  LLVMFastMathAllowContract = (1 << 5),
+  LLVMFastMathApproxFunc = (1 << 6),
+  LLVMFastMathNone = 0,
+  LLVMFastMathAll = -1,
```
```suggestion
  LLVMFastMathAll = LLVMFastMathAllowReassoc | LLVMFastMathNoNaNs | ...,
```

nikic commented on llvm/include/llvm-c/Core.h:
```diff
@@ -483,6 +483,26 @@ typedef enum {
+
+/**
+ * Flags to indicate what fast-math-style optimizations are allowed
+ * on operations
```
```suggestion
 * on operations.
```
Here and elsewhere.

nikic commented on llvm/include/llvm-c/Core.h:
```diff
@@ -4075,6 +4095,32 @@ LLVMBool LLVMGetNNeg(LLVMValueRef NonNegInst);
+ * Only valid on floating point instructions
+ * @see LLVMGetCanUseFastMathFlags
+ */
+LLVMFastMathFlags LLVMGetFastMathFlags(LLVMValueRef FPMathInst);
```
Missing newline.

nikic commented on llvm/include/llvm-c/Core.h:
```diff
@@ -4075,6 +4095,32 @@ LLVMBool LLVMGetNNeg(LLVMValueRef NonNegInst);
+ *
+ * Will return true for floating point arithmetic instructions, and for select,
+ * phi, and call instructions whose type is a floating point type, or a vector
+ * or array thereof See https://llvm.org/docs/LangRef.html#fast-math-flags
```
```suggestion
 * or array thereof. See https://llvm.org/docs/LangRef.html#fast-math-flags
```

nikic commented on llvm/lib/IR/Core.cpp:
```diff
@@ -3319,6 +3319,52 @@ void LLVMSetArgOperand(LLVMValueRef Funclet, unsigned i, LLVMValueRef value) {
+  FastMathFlags NewFMF;
+  // First, check if all bits are set
+  // If not, check each one explicitly
+  if (FMF == static_cast<LLVMFastMathFlags>(LLVMFastMathAll))
```
Can drop this after defining LLVMFastMathAll correctly.

nikic commented on llvm/lib/IR/Core.cpp:
```diff
@@ -3319,6 +3319,52 @@ void LLVMSetArgOperand(LLVMValueRef Funclet, unsigned i, LLVMValueRef value) {
+}
+
+static LLVMFastMathFlags mapToLLVMFastMathFlags(FastMathFlags FMF) {
+
```
Stray newline.

nikic commented on llvm/test/Bindings/llvm-c/float_ops.ll:
```diff
@@ -0,0 +1,155 @@
+;
+source_filename = "/test/Bindings/float_ops.ll"
+target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-apple-macosx10.11.0"
```
Drop the triple.

nikic commented on llvm/tools/llvm-c-test/echo.cpp:
```diff
@@ -770,8 +770,19 @@ struct FunCloner {
+        LLVMValueRef Else = CloneValue(LLVMGetOperand(Src, 2));
+        Dst = LLVMBuildSelect(Builder, If, Then, Else, Name);
+        if (LLVMGetCanUseFastMathFlags(Src))
+          LLVMSetFastMathFlags(Dst, LLVMGetFastMathFlags(Src));
```
Rather than adding this to specific cases (and having to list out all the FP instructions), can we do this once after the switch?

Benjins committed:
Address PR feedback

Benjins commented:
Pushed a new commit that addresses review feedback:
 - Periods in documentation and formatting
 - Changing `LLVMFastMathAll` to be a bitwise or of all the other flags, rather than `-1`
 - Move `LLVMSetFastMathFlags` to after switch statement in `echo.cpp` for non-phi instructions
 - No target triple in `float_ops.ll`

nikic reviewed (approved):
LGTM, though I wonder whether `LLVMGetCanUseFastMathFlags` should be just `LLVMCanUseFastMathFlags`.

We don't seem to have anything using "LLVMCan" right now, but we do have "LLVMHas" and "LLVMIs".

Benjins committed:
Rename LLVMGetCanUseFastMathFlags to LLVMCanValueUseFastMathFlags
